Kernel/Devices/HID/VirtIO/Input.cpp

```diff
@@ -0,0 +1,497 @@
+    transport_entity().read_config_atomic([&]() {
+        auto size = transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, size));
+        if (size == 0)
+            return;
+
+        VERIFY(size <= sizeof(VirtIOInputConfig::u.string));
+
+        char* name_chars = nullptr;
+        name = MUST(KString::try_create_uninitialized(size, name_chars));
+
+        for (size_t i = 0; i < size; i++)
+            name_chars[i] = static_cast<char>(transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, u.string) + i));
+    });
+
+    if (name)
+        dbgln("VirtIO::Input: Device name: {}", name);
+
+    Optional<VirtIOInputConfig::DevIDs> maybe_dev_ids;
+
+    transport_entity().config_write8(*cfg, offsetof(VirtIOInputConfig, select), to_underlying(VirtIOInputConfig::Select::IDDevIDs));
+    transport_entity().read_config_atomic([&]() {
+        auto size = transport_entity().config_read8(*cfg, offsetof(VirtIOInputConfig, size));
+        if (size == 0)
+            return;
+
+        VERIFY(size == sizeof(VirtIOInputConfig::u.ids));
+
+        maybe_dev_ids = VirtIOInputConfig::DevIDs {
+            .bustype = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.bustype))),
+            .vendor = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.vendor))),
+            .product = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.product))),
+            .version = bit_cast<LittleEndian<u16>>(transport_entity().config_read16(*cfg, offsetof(VirtIOInputConfig, u.ids.version))),
+        };
+    });
+
+    if (maybe_dev_ids.has_value()) {
+        auto dev_ids = maybe_dev_ids.value();
+
+        dbgln("VirtIO::Input: DevIDs: bustype={:#x}, vendor={:#x}, product={:#x}, version={:#x}", dev_ids.bustype, dev_ids.vendor, dev_ids.product, dev_ids.version);
+
+        // Use the dev_ids to determine if the device is a QEMU mouse or keyboard, as we don't support devices sending both types of input events.
+        if (dev_ids.bustype == 0x6 && dev_ids.vendor == 0x627) {
+            if (dev_ids.product == 0x1) {
+                // QEMU Virtio Keyboard
+                m_keyboard_device = KeyboardDevice::try_to_initialize().release_value_but_fixme_should_propagate_errors();
+                HIDManagement::the().attach_standalone_hid_device(*m_keyboard_device);
+            } else if (dev_ids.product == 0x2 || dev_ids.product == 0x3) {
+                // QEMU Virtio Mouse/Tablet
+                m_mouse_device = MouseDevice::try_to_initialize().release_value_but_fixme_should_propagate_errors();
+                HIDManagement::the().attach_standalone_hid_device(*m_mouse_device);
+            } else {
+                dbgln("Virtio::Input: Unknown product ID");
+                VERIFY_NOT_REACHED();
```

Comment:
There sadly is no nice way to ignore unknown devids, since the caller of `initialize_virtio_resources` uses `MUST.` We also would have to leave the Input device in a half-initialized state, which doesn't seem ideal.
So I would rather wait for #24210 to be merged, since I don't think we should be using `VERIFY_NOT_REACHED` here.

Reply:
I removed the devid matching code now.
Only the mouse will reliably work until #24210 is merged.

The keyboard actually works with the way it is added in run.py because it is listed first, and QEMU apparently assigns PCI addresses based on the order of the command line arguments.
As a result, the keyboard is located at "/dev/input/keyboard/0", which is where WindowServer expects it to be.